routing_common/bicycle_model.cpp

```diff
@@ -28,6 +28,10 @@ namespace
 // As result of such heuristic road is not totally the shortest, but it avoids non bicycle roads, which were
 // not marked as "hwtag=nobicycle" in OSM.
 
+InOutCitySpeedKMpH constexpr kFormerModelMaxSpeed = {
```

Comment:
почему она тут только одна? тут же должен быть вектор в общем в случае

Reply:
Идея в том, что это максимум с предыдущих времен. Когда мы вычисляем max speed текущей модели (`m_maxModelSpeed`), мы смотрим, чтоб он `m_maxModelSpeed` был не меньше, чем максимальный max speed с прежних времен.

Reply:
предлагаю написать об этом комментарий

Reply:
и вообще написать комментарий - зачем мы храним предыдущие скоростим, можно с маленьким примером, аля:
```
      C
     /   \
   /       \
A ----- B
For A* is essential that time(A, B) <= time(A, C) + time(C, B). When we use LeapsOnly mode we calculate some weights of roads during map generation. 
// TODO
```
Пока писал, снова перестал понимать, почему это падает, ведь в LeapsOnly mode нет эвристики, там дейкстра, надо будет разобраться и дописать коммент

Reply:
> Пока писал, снова перестал понимать, почему это падает, ведь в LeapsOnly mode нет эвристики, там дейкстра, надо будет разобраться и дописать коммент

Насколько я понимаю, в LeapsOnly односторонний A*, т.к. нам было тяжело писать двусторонний. Смотри.
В `IndexRouter::CalculateSubroute()` в случае режима LeapsOnly вызывается `FindPath()` с графом `IndexGraphStarter`. А в нем `HeuristicCostEstimate()` методы возвращают валидные значения. Ну а код `FindPath()` - это и есть A*.

```c++
  template <typename Vertex, typename Edge, typename Weight>
  RouterResultCode FindPath(
      typename AStarAlgorithm<Vertex, Edge, Weight>::Params & params, std::set<NumMwmId> const & mwmIds,
      RoutingResult<Vertex, Weight> & routingResult, WorldGraphMode mode) const
  {
    AStarAlgorithm<Vertex, Edge, Weight> algorithm;
    if (mode == WorldGraphMode::LeapsOnly)
    {
      return ConvertTransitResult(mwmIds,
                                  ConvertResult<Vertex, Edge, Weight>(algorithm.FindPath(params, routingResult)));
    }
    return ConvertTransitResult(
        mwmIds, ConvertResult<Vertex, Edge, Weight>(algorithm.FindPathBidirectional(params, routingResult)));
  }
```

Reply:
> и вообще написать комментарий - зачем мы храним предыдущие скоростим, можно с маленьким примером, аля:

Тогда придется писать комментарий около каждой константы. А их столько сколько моделей. Сейчас появился метод `void VehicleModel::TuneMaxModelSpeed(InOutCitySpeedKMpH const & formerModelMaxSpeed)` давай напишем подробный коммент с примером около него.

Reply:
Предлагаю вот так:
```c++
  /// \brief Increases |m_maxModelSpeed| if it's less than |formerModelMaxSpeed|.
  /// \note It's necessary because speeds in car model, pedestrian model and so on
  /// are changed during application life. Based on this speeds cross_mwm sections
  /// are generated. So the old speeds in old models affect old mwms. But old mwms
  /// may be used in a new version of the app. Section cross_mwm is used for cross mwm
  /// routing in LeapsOnly mode. For calculation cross mwm route A* with a heuristic
  /// based on the quickest time needed to get from one point to another is used.
  /// For calculation the quickest time |m_maxModelSpeed| is used.
  /// Because of it's important that |m_maxModelSpeed| is maximum speed for all period
  /// when mwm supported.
  void TuneMaxModelSpeed(InOutCitySpeedKMpH const & formerModelMaxSpeed);
```
И около каждой константы с прежними скоростями - написал, где написано подробно.

Reply:
да, все, я перепутал generator и LeapsOnly

Reply:
```
1) We use maxspeed for calculate time between two points (dist / modelMaxSpeed)
2) We use maxspeed for calculate time on the fastest road (highway-motorway)
    C
  /   \
 /     \
A ----- B
For A* is essential that time(A, B) <= time(A, C) + time(C, B). When we use LeapsOnly mode we calculate weights of leaps (path from enter to exit of mwm) during map generation. 
Consider A is enter and B is exit, and A -> B is very small leap consists of 1 road of type highway-motorway (we use the model's maxSpeed on such type of roads). So, assume that we calculated the weight of A -> B with maxspeed = 100mps and saved this data to mwm, and assume we update the code and increase the maxspeed to 110 mps, so it could happened that exists point C so: time(A, B) > time(A, C) + time(C, B). Because time(A, B) was calculated in the past with maxspeed 100 mps, and time(A, C) and time(C, B) calculates with new maxspeed = 110 mps.
```

Reply:
давай этот комментарий добавим для car model

Reply:
> > Пока писал, снова перестал понимать, почему это падает, ведь в LeapsOnly mode нет эвристики, там дейкстра, надо будет разобраться и дописать коммент
> 
> Насколько я понимаю, в LeapsOnly односторонний A*, т.к. нам было тяжело писать двусторонний. Смотри.
> В `IndexRouter::CalculateSubroute()` в случае режима LeapsOnly вызывается `FindPath()` с графом `IndexGraphStarter`. А в нем `HeuristicCostEstimate()` методы возвращают валидные значения. Ну а код `FindPath()` - это и есть A*.
> 
> ```c++
>   template <typename Vertex, typename Edge, typename Weight>
>   RouterResultCode FindPath(
>       typename AStarAlgorithm<Vertex, Edge, Weight>::Params & params, std::set<NumMwmId> const & mwmIds,
>       RoutingResult<Vertex, Weight> & routingResult, WorldGraphMode mode) const
>   {
>     AStarAlgorithm<Vertex, Edge, Weight> algorithm;
>     if (mode == WorldGraphMode::LeapsOnly)
>     {
>       return ConvertTransitResult(mwmIds,
>                                   ConvertResult<Vertex, Edge, Weight>(algorithm.FindPath(params, routingResult)));
>     }
>     return ConvertTransitResult(
>         mwmIds, ConvertResult<Vertex, Edge, Weight>(algorithm.FindPathBidirectional(params, routingResult)));
>   }
> ```

Судя по коду, у нас вообще нигде не используется Дейкстра. Для LeapsOnly вызывается обычный A* (FindPath, а не FindPathBidirectional).

Reply:
ping about comment

Reply:
Добавил коммент к методу `VehicleModel::TuneMaxModelSpeed`:
```c++
  /// \brief Increases |m_maxModelSpeed| if it's less than |formerModelMaxSpeed|.
  /// \note
  ///  1) We use maxspeed for calculation time between two points (dist / modelMaxSpeed)
  ///  2) We use maxspeed for calculation time on the fastest road (highway-motorway)
  ///      C
  ///    /   \
  ///   /     \
  ///  A ----- B
  ///      For A* is essential that time(A, B) <= time(A, C) + time(C, B). When we use LeapsOnly
  ///  mode we calculate weights of leaps (path from enter to exit of mwm) during map generation.
  ///  Let's consider A is enter and B is exit, and A -> B is very small leap consists of 1 road of type
  ///  highway-motorway. So, let us assume that the weight of A -> B is calculated with maxspeed = 100mps
  ///  and saved this data to mwm, and let us assume we update the code and increase the maxspeed to 110 mps,
  ///  so it could happened that exists point C so: time(A, B) > time(A, C) + time(C, B). Because time(A, B)
  ///  was calculated in the past with maxspeed 100 mps, and time(A, C) and time(C, B) are calculated with new
  ///  maxspeed = 110 mps.
  void TuneMaxModelSpeed(InOutCitySpeedKMpH const & formerModelMaxSpeed);
```